### PR TITLE
TESTS: Use File Based Discovery in REST Tests

### PR DESCRIPTION
**Pull request by original-brownbear**
* For 7.0 use file based discovery in REST tests
* Relates #33675

### CODE REVIEW

**Comment by original-brownbear** [comment]
@DaveCTurner I hope this is what we want here :)
Also, should I do anything about pre-7.x here? (I guess that would require using the file based discovery plugin in older versions?)

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy [hunk 1]
```diff
@@ -703,6 +707,22 @@ class ClusterFormationTasks {
                     }
                 }
             }
+
+            nodes.forEach {node ->
+                if (node.nodeVersion.onOrAfter("7.0.0-SNAPSHOT")) {
+                    Collection<String> unicastHosts = new HashSet<>()
+                    nodes.forEach { otherNode ->
```
Comment (DaveCTurner) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy]: I think this should only be the master-eligible nodes.

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
File-based discovery is available without a plugin in `6.x` so I think once 6.5.0 is released we can do this (in `master`) without worrying about versions. We should wait.

Also, with this change I think the whole concept of a "seed" node can go away.

Also, with this change I think we can undo the changes in #27344 and go back to a more reasonable setting for `minimum_master_nodes`.

**Comment by DaveCTurner** [comment]
> Also, should I do anything about pre-7.x here? (I guess that would require using the file based discovery plugin in older versions?)

I'm not sure - how easy is it to install a plugin for these tests? If it's easy, I think we should try and make an equivalent change to `6.x` that does indeed involve installing the `discovery-file` plugin before starting these nodes.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy (hunk 1 above)
Comment (original-brownbear) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy]: I thought the same, but then I found this code https://github.com/elastic/elasticsearch/blob/master/buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterConfiguration.groovy#L71 => aren't all nodes master-eligible nodes in these tests?

**Comment by original-brownbear** [comment]
@DaveCTurner I think there is logic that helps with installing plugins in this code but I would need to research this a little to see how tricky it is to actually use it.
Are you suggesting otherwise generally blocking the work here until `6.5.0` is out?

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy (hunk 1 above)
Comment (original-brownbear) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy]: Ah now I read #27344, so I guess until we change that all nodes are master eligible indeed?

Comment (DaveCTurner) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy]: You are correct: now that I look more closely it seems that all nodes in these tests are master-eligible. Don't mind me...

**Comment by DaveCTurner** [comment]
> Are you suggesting otherwise generally blocking the work here until 6.5.0 is out?

Yes.

**Comment by DaveCTurner** [comment]
> > Are you suggesting otherwise generally blocking the work here until 6.5.0 is out?

> Yes.

Wrong again! I thought we tested `master` against `6.4` as well as `6.x` but apparently not. This PR can proceed, sorry.

**Comment by original-brownbear** [comment]
@DaveCTurner no worries :) I'm back on it then

**Comment by original-brownbear** [comment]
@DaveCTurner alright, that was a confusing one (sorry for missing that `6.x` also supports the file based discovery out of the box :D) => Removed the condition for the version now. Take another look when you have a sec :)

**Comment by original-brownbear** [comment]
@DaveCTurner argh :D see Jenkins log ... we do run against 6.4.x-snapshot => back to waiting it out here? :)

**Comment by original-brownbear** [comment]
@DaveCTurner alright back to plan A as discussed: Made it only use file based discovery for 6.5+ now. Should be good for another review :)

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy [hunk 2]
```diff
@@ -347,9 +347,16 @@ class ClusterFormationTasks {
 
         Task writeConfig = project.tasks.create(name: name, type: DefaultTask, dependsOn: setup)
         writeConfig.doFirst {
-            String unicastTransportUri = node.config.unicastTransportUri(seedNode, node, project.ant)
-            if (unicastTransportUri != null) {
-                esConfig['discovery.zen.ping.unicast.hosts'] = "\"${unicastTransportUri}\""
+            if (node.nodeVersion.onOrAfter("6.5.0-SNAPSHOT")) {
+                // Don't force discovery provider if one is set by the test cluster specs already
+                if (esConfig.containsKey('discovery.zen.hosts_provider') == false) {
+                    esConfig['discovery.zen.hosts_provider'] = 'file'
```
Comment (DaveCTurner) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy]: I think we should also set `discovery.zen.ping.unicast.hosts: []` or else we will attempt to ping ports `9300-9304` too.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy [hunk 3]
```diff
@@ -347,9 +347,16 @@ class ClusterFormationTasks {
...
+                }
+            } else {
+                String unicastTransportUri = node.config.unicastTransportUri(seedNode, node, project.ant)
+                if (unicastTransportUri != null) {
+                    esConfig['discovery.zen.ping.unicast.hosts'] = "\"${unicastTransportUri}\""
```
Comment (DaveCTurner) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy]: I think if `unicastTransportUri == null` we should also set `discovery.zen.ping.unicast.hosts: []` or else we will attempt to ping ports `9300-9304` too (this is unrelated to this change, I just noticed it).

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy [hunk 4]
```diff
@@ -681,6 +688,20 @@ class ClusterFormationTasks {
     static Task configureWaitTask(String name, Project project, List<NodeInfo> nodes, List<Task> startTasks, int waitSeconds) {
         Task wait = project.tasks.create(name: name, dependsOn: startTasks)
         wait.doLast {
+
+            nodes.forEach { node ->
+                Collection<String> unicastHosts = new HashSet<>()
+                nodes.forEach { otherNode ->
```
Comment (DaveCTurner) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy]: Does this need to be nested? Could we calculate it once up-front and then write all the hosts files?

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy [hunk 5]
```diff
@@ -703,6 +724,7 @@ class ClusterFormationTasks {
                     }
                 }
             }
+
```
Comment (DaveCTurner) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy]: You could make this PR 4% smaller by removing this line again ;)

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
Neat! Can we lose the start-one-node-then-start-the-rest pattern for 6.5.0+ clusters? I asked for a handful of minor changes too.

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
(copied from previous review): Also, with this change I think we can undo the changes in #27344 and go back to a more reasonable setting for minimum_master_nodes.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy (hunk 4 above)
Comment (original-brownbear) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy]: Wouldn't we want to not have self-references in these (I guess it doesn't matter), but in that case we actually need additional code to account for the one self-reference that will return `null` in that loop.
I just figured this was the shortest solution? :)

**Comment by original-brownbear** [comment]
@DaveCTurner 

> Can we lose the start-one-node-then-start-the-rest pattern for 6.5.0+ clusters?

Sure we could, but do we actually want to? I'm not sure doing that has any concrete benefit but forces us to add more code for handling the different paths with and without `seedNode`?

> Also, with this change I think we can undo the changes in #27344 and go back to a more reasonable setting for minimum_master_nodes.

Want to do this here or in a follow up?

**Comment by original-brownbear** [comment]
@DaveCTurner minor changes are in :) Let me know how to proceed on the other points.

**Comment by DaveCTurner** [comment]
> > Can we lose the start-one-node-then-start-the-rest pattern for 6.5.0+ clusters?
> 
> Sure we could, but do we actually want to? I'm not sure doing that has any concrete benefit but forces us to add more code for handling the different paths with and without `seedNode`?

Having the paths determined by the version makes it clearer that `seedNode` is only important for pre-6.5.0 clusters. Otherwise it'll catch someone out in future, and we will have to engage our brains again in the future to think about removing it in v8.0. It's also a step towards starting the nodes with the right value for `discovery.initial_state_timeout` which gives us more realistic tests. If it's super-messy then we can live without it, but I think it'd be not so bad.

> > Also, with this change I think we can undo the changes in #27344 and go back to a more reasonable setting for minimum_master_nodes.
> 
> Want to do this here or in a follow up?

Unless it's super-messy let's do it here.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy (hunk 4 above)
Comment (DaveCTurner) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy]: Ok, more seed node weirdness. I think you should up-front call `unicastTransportUri(node, null, project.ant)` in order to read the transport address of every node, and then put all of them into all of the `unicast_hosts.txt` files. It's normal for all the nodes to have the same list of unicast hosts, even though one of them refers to itself.

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy [hunk 6]
```diff
@@ -347,9 +347,19 @@ class ClusterFormationTasks {
...
         Task writeConfig = project.tasks.create(name: name, type: DefaultTask, dependsOn: setup)
         writeConfig.doFirst {
-            String unicastTransportUri = node.config.unicastTransportUri(seedNode, node, project.ant)
-            if (unicastTransportUri != null) {
-                esConfig['discovery.zen.ping.unicast.hosts'] = "\"${unicastTransportUri}\""
+            if (node.nodeVersion.onOrAfter("6.5.0-SNAPSHOT")) {
+                // Don't force discovery provider if one is set by the test cluster specs already
+                if (esConfig.containsKey('discovery.zen.hosts_provider') == false) {
+                    esConfig['discovery.zen.hosts_provider'] = 'file'
+                    esConfig['discovery.zen.ping.unicast.hosts'] = []
```
Comment (DaveCTurner) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy]: I'm concerned that this should also be set to `[]` even if `discovery.zen.hosts_provider` is set externally. The places where I can see it being set externally are `azure`, `ec2` and `gce` and in all these cases we don't want to be probing local ports `9300-9304` too.

Comment (original-brownbear) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterFormationTasks.groovy]: Makes sense will do :)

**Comment by original-brownbear** [comment]
@DaveCTurner alright all handled I think:

The waiting for the seed node was easy to fix functionally in https://github.com/elastic/elasticsearch/pull/34560/commits/391a8a6afd5619bf770ad28edc2e3a02f787d420 (just don't add the task dependency to new version nodes). I didn't refactor the code for it a great deal because of the last point below

The precomputing the unicast hosts and setting `[]` for the unicast hosts in the missing spots is in https://github.com/elastic/elasticsearch/pull/34560/commits/2052e0c91b5f0bf762c0f63def72b2d0b7936cf9

The situation with the minimum master nodes setting I don't think I can easily fix here. For the BwC nodes this will still spell trouble because we're still using the old way of setting the unicast hosts to the first node for those and nothing has changed for them? => I didn't make any changes to that and also figured there wasn't too much point in a larger code change to separate the paths for older than `6.5` and newer than `6.5` until we had a plan for that. Maybe add a new issue for that and look for a clean refactoring and fix for the pre-`6.5` situation there?

**Comment by original-brownbear** [comment]
Jenkins test this

**Comment by original-brownbear** [comment]
@DaveCTurner looks like GH finally refreshed :) This should be good for another review now.

**Comment by original-brownbear** [comment]
Jenkins test this

**Comment by original-brownbear** [comment]
Jenkins test this

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterConfiguration.groovy [hunk 7]
```diff
@@ -131,7 +137,7 @@ class ClusterConfiguration {
      * condition is executed.
      */
     @Input
-    int nodeStartupWaitSeconds = 30
+    int nodeStartupWaitSeconds = 120
```
Comment (DaveCTurner) [inline buildSrc/src/main/groovy/org/elasticsearch/gradle/test/ClusterConfiguration.groovy]: I'd rather not do this - I think there's something wrong with the discovery plugin tests.

**Review by DaveCTurner — COMMENTED** [review]
I'm happy with this (except the longer timeout, see comment). However CI is not. We looked into it together and suspected that there's something wrong with how the discovery plugin tests work. If so, and if it fixes it, I'd be ok with reinstating the `discovery.zen.ping.unicast.hosts` lists in those cases and then following up with a proper fix as a separate PR.

**Comment by original-brownbear** [comment]
@DaveCTurner it seems the normal discovery mechanism interfering (or rather it not doing so) wasn't the issue here.
We're making all the nodes return `null` for their transport endpoint in https://github.com/elastic/elasticsearch/blob/ca6808e55de5b9b5d0daa074c1512366fcb57fde/plugins/discovery-ec2/qa/amazon-ec2/build.gradle#L59 => still at it.
Looking into the changed dependency order now.

**Review by DaveCTurner — APPROVED** [review]
LGTM
